src/psyclone/psyir/transformations/inline_trans.py

```diff
@@ -413,62 +791,139 @@ def validate(self, node, options=None):
+                        f"because it accesses variable '{sym.name}' and this "
+                        f"cannot be found in any of the containers directly "
+                        f"imported into its symbol table.")
+            else:
+                if sym.name not in routine_table:
                     raise TransformationError(
                         f"Routine '{routine.name}' cannot be inlined because "
-                        f"it accesses variable '{ref.symbol.name}' from its "
+                        f"it accesses variable '{sym.name}' from its "
                         f"parent container.")
 
-        # Check that the shape of any dummy array arguments are the same as
+        # Check that the shape of any formal array arguments are the same as
         # those at the call site.
         visitor = FortranWriter()
-        for dummy_arg, actual_arg in zip(routine_table.argument_list,
-                                         node.children):
-            # TODO #1799 this really needs the `datatype` method that is
-            # planned for all reference nodes.
-            if isinstance(actual_arg, ArrayReference):
-                rank = 0
-                for idx, expr in enumerate(actual_arg.indices):
-                    if isinstance(expr, Range):
-                        rank += 1
-                        # TODO #924 add a method to Range that returns the
-                        # PSyIR expression for the number of elements it
-                        # contains. This can then be compared with the shape of
-                        # the dummy argument.
-                        if not actual_arg.is_full_range(idx):
-                            # It's OK to use the loop variable in the lambda
-                            # definition because if we get to this point then
-                            # we're going to quit the loop.
-                            # pylint: disable=cell-var-from-loop
-                            raise TransformationError(LazyString(
-                                lambda: f"Cannot inline routine "
-                                f"'{routine.name}' because argument "
-                                f"'{visitor(actual_arg)}' is "
-                                f"an array subsection (TODO #924)."))
-                if isinstance(dummy_arg.datatype, ScalarType):
-                    dummy_rank = 0
-                else:
-                    dummy_rank = len(dummy_arg.datatype.shape)
-                if rank != dummy_rank:
-                    # It's OK to use the loop variable in the lambda definition
-                    # because if we get to this point then we're going to quit
-                    # the loop.
-                    # pylint: disable=cell-var-from-loop
-                    raise TransformationError(LazyString(
+        for formal_arg, actual_arg in zip(routine_table.argument_list,
+                                          node.children):
```

Comment:
Should we check that their len() matches and produce a TransformationError if not?  Because zip is happy to ignore elements if one list is longer than the other.

Reply:
Yes, we should - another good spot. That found an error in one of the tests too.